roll/__init__.py

```diff
@@ -325,9 +339,14 @@ def data_received(self, data: bytes):
             self.response.status = HTTPStatus.BAD_REQUEST
             self.response.body = b'Unparsable request'
             self.write()
-
-    def connection_made(self, transport):
-        self.writer = transport
+        except HttpParserUpgrade:
+            # https://tools.ietf.org/html/rfc2616#page-144
+            # We precise the upgrade type requested on the request itself.
+            # It's up to the handler to act on it or not.
+            # This header should be parsed and maybe validated if we want
+            # to handle complex declarations, such as :
+            # Upgrade: HTTP/2.0, SHTTP/1.3, IRC/6.9, RTA/x11
+            self.request.upgrade = self.request.headers['UPGRADE'].lower()
```

Comment:
Not sure about the pattern yet: seems we are mixing a bit the Protocol and the Request responsibilities.
I feel like the big picture is:
HttpProtocol => Upgrade received => close http response => set_protocol(WSProtocol).

What about a pair session so we can tackle this down at four hands? :)

Reply:
Pair programming sounds good. Whenever you want